src/cls/rbd/cls_rbd_types.h

```diff
@@ -216,6 +217,140 @@ struct GroupSpec {
 
+enum SnapshotNamespaceType {
+  SNAPSHOT_NAMESPACE_TYPE_USER = 0,
+  SNAPSHOT_NAMESPACE_TYPE_GROUP = 1
+};
+
+struct UserSnapshotNamespace {
+  static const uint32_t SNAPSHOT_NAMESPACE_TYPE = SNAPSHOT_NAMESPACE_TYPE_USER;
+
+  UserSnapshotNamespace() {}
+
+  void encode(bufferlist& bl) const {}
+  void decode(bufferlist::iterator& it) {}
+
+  inline bool operator==(const UserSnapshotNamespace& usn) const {
+    return true;
+  }
+
+};
+
+WRITE_CLASS_ENCODER(UserSnapshotNamespace);
+
+std::ostream& operator<<(std::ostream& os, const UserSnapshotNamespace& ns);
+
+struct GroupSnapshotNamespace {
+  static const uint32_t SNAPSHOT_NAMESPACE_TYPE = SNAPSHOT_NAMESPACE_TYPE_GROUP;
+
+  GroupSnapshotNamespace() {}
+
+  GroupSnapshotNamespace(int64_t _group_pool,
+			 const string &_group_id,
+			 const string &_snapshot_id) :group_pool(_group_pool),
+						      group_id(_group_id),
+						      snapshot_id(_snapshot_id) {}
+
+  int64_t group_pool;
+  string group_id;
+  string snapshot_id;
+
+  void encode(bufferlist& bl) const {
+    ::encode(group_pool, bl);
+    ::encode(group_id, bl);
+    ::encode(snapshot_id, bl);
+  }
+
+  void decode(bufferlist::iterator& it) {
+    ::decode(group_pool, it);
+    ::decode(group_id, it);
+    ::decode(snapshot_id, it);
+  }
+
+  inline bool operator==(const GroupSnapshotNamespace& gsn) const {
+    return group_pool == gsn.group_pool &&
+	   group_id == gsn.group_id &&
+	   snapshot_id == gsn.snapshot_id;
+  }
+
+};
+
+WRITE_CLASS_ENCODER(GroupSnapshotNamespace);
+
+std::ostream& operator<<(std::ostream& os, const GroupSnapshotNamespace& ns);
+
+struct UnknownSnapshotNamespace {
+  static const uint32_t SNAPSHOT_NAMESPACE_TYPE = static_cast<uint32_t>(-1);
+
+  UnknownSnapshotNamespace() {}
+
+  void encode(bufferlist& bl) const {}
+  void decode(bufferlist::iterator& it) {}
+  inline bool operator==(const UnknownSnapshotNamespace& gsn) const {
+    return true;
+  }
+};
+
+WRITE_CLASS_ENCODER(UnknownSnapshotNamespace);
+
+std::ostream& operator<<(std::ostream& os, const UnknownSnapshotNamespace& ns);
+
+typedef boost::variant<UserSnapshotNamespace, GroupSnapshotNamespace, UnknownSnapshotNamespace> SnapshotNamespace;
+
+class EncodeSnapshotTypeVisitor : public boost::static_visitor<void> {
+public:
+  explicit EncodeSnapshotTypeVisitor(bufferlist &bl) : m_bl(bl) {
+  }
+
+  template <typename T>
+  inline void operator()(const T& t) const {
+    ::encode(static_cast<uint32_t>(T::SNAPSHOT_NAMESPACE_TYPE), m_bl);
+    t.encode(m_bl);
+  }
+
+private:
+  bufferlist &m_bl;
+};
+
+class DecodeSnapshotTypeVisitor : public boost::static_visitor<void> {
+public:
+  DecodeSnapshotTypeVisitor(bufferlist::iterator &iter)
+    : m_iter(iter) {
+  }
+
+  template <typename T>
+  inline void operator()(T& t) const {
+    t.decode(m_iter);
+  }
+private:
+  bufferlist::iterator &m_iter;
+};
+
+inline void encode(const SnapshotNamespace &c, bufferlist &bl, uint64_t features=0)
+{
+  ENCODE_DUMP_PRE();
```

Comment:
I'm not sure I understand this comment. Are you talking about ENCODE_DUMP_PRE(); specifically? I see that WRITE_CLASS_ENCODER is used in order to create encode and decode functions for MirrorImageStatus. WRITE_CLASS_ENCODER adds ENCODE_DUMP_PRE and ENCODE_DUMP post inside encode function it generates.